### DISCUSSION TITLE
fetch stream

### DISCUSSION THREAD

**Discussion by wcastand** [expo/expo#21710] [Proposals and Ideas]
Hi, is there any plan on adding support for stream in the fetch api or a polyfill?

using yoga-server and urql and i can't since it requires streams on the fetch api.

In theory, there is a polyfill but it's not up to date and seems to be abandoned. https://www.npmjs.com/package/react-native-fetch-api?activeTab=readme

**Comment by wcastand** [comment]
do we actually need to bump or like the discord no one really cares? :)


EDIT: Sorry, just a bit frustrated of never getting any answer here or on discord, most discussion seems to be ignored completely and all the issues i find are always closed by expo-bot even if the issue is never seen by someone from expo team or even fixed. 

It's not always possible to offer a repo but that doesn't mean the issue doesn't exist and altho i try to help i can't spend 2 weeks a bug fix during work hours (guessing it's part of the expo team to handle the investigation, i don't ask my client to investigate their own problem on my app)

I just liked to know hours where expo team is available to help and see more interaction here or on discord because in more than a year of using expo i got help/answer once on discord and discussions and i'd rather not have to make tickets for every single things and annoy the support team when there is discord an github repo that are suppose to be there for this :/

**Comment by KishiTheMechanic** [comment]
We need the stream on TypeScript React Native.
From my quick research, I couldn't find the right repo so far.

How can we solve it?

**Comment by heethjain21** [comment]
The fetch package url which you shared works

Even though it's a bit old. I got it working a few days ago.

Can use this for reference: https://stackoverflow.com/a/77089139

Also check out the comment for if you're having issues with Android (it doesn't work in debug mode)

**Reply by robinsadeghpour** [reply]
I cannot call res.body.getReader() it always says `Error: [TypeError: This stream has already been locked for exclusive reading by another reader]`. Have you encountered this issue as well?

**Reply by heethjain21** [reply]
I don't remember, it has been quite a while. 

here is my code if you want to use for reference:

```
      fetch(`${API_URL}`, {
        method: 'POST',
        headers: {
          // as per needed
        },
        body: JSON.stringify({
          // as per needed
        }),
        reactNative: { textStreaming: true },
        signal: abortController.signal, // Pass the AbortSignal to the fetch request
      })
        .then(response => response.body)
        .then(async (stream) => {

          const reader = stream.getReader();
          const decoder = new TextDecoder();

          // fetch is complete and chat bot started typing

          const processStream = async () => {
            const processChunk = async () => {

              const { done, value } = await reader.read();

              if (done) {
                //do something
                return;
              }

              const text = decoder.decode(value); // Convert the binary data to text
              resultText += text;

              setTimeout(processChunk); // Delay before processing the next chunk
            };
            processChunk();
          };
          processStream();
        }).catch((error) => {
          //do something
        }).finally(async () => {
          // do something
        });
```

abortController is for stopping the fetch api stream response in between. So with a function something like this

```
  const [abortController, setAbortController] = useState(new AbortController()); // not sure if using state here is the best solution. I had to do a bit experimenting and finally settle on this

  const handleStop = () => {
    try {
      if (abortController) {
        abortController.abort(); // Abort the fetch request
        setAbortController(new AbortController()); // set new controller for next fetch req
        // do other things
      }
    }
    catch (error) {
      console.error(error);
    }

  };
```

**Reply by ybentz** [reply]
@heethjain21 do streams work properly on Android for you? I got everything working with the polyfills and iOS works great but on Android what I'm seeing is a very long delay in first stream chunk and then it just "unloads" all the chunks at the same time when the stream closes. It's almost like there's a very big buffer that accumulates all the stream chunks and then flushes them all at the same time. I don't think it could have something to do with the polyfills because those are JS and the issue is only on Android so probably in native code.

I saw the comment about Android but I don't even have that Flipper file (I'm using Expo v50).

**Reply by heethjain21** [reply]
@ybentz I was on a bare react native project (without expo), so the Flipper file was related to that.

And it worked fine on Android for me.

What type of build/project are you on?

I can share the specific/particular Android files if needed

**Reply by ybentz** [reply]
@heethjain21 I'm on Expo 50 and it's a managed project. Did you mean anything Android specific? I'm not doing anything special on my end, just using Expo and building using EAS either locally or using their cloud service.

**Reply by heethjain21** [reply]
@ybentz apologies, for some reason I missed it.

Yes, I was on a bare react native project which means it "Flipper" file was in the /android directory.

So android specific files, which would not be available on Expo managed project.

**Comment by tao-qian** [comment]
We ran into the same problem and ended up using [react-native-sse](https://github.com/binaryminds/react-native-sse) instead. 

Here is an [example](https://github.com/tao-qian/react-native-chat-streaming-example/blob/main/app/index.tsx) for streaming responses from the openai API.

**Comment by Kudo** [comment]
FWIW, The upcoming Expo SDK 52 will have an `expo/fetch` to support streaming.
reference code:

https://github.com/expo/expo/blob/89397f4ebc4776d6bb7fdb0053c2218435e7e9e5/apps/test-suite/tests/Fetch.ts#L1
https://github.com/expo/expo/blob/89397f4ebc4776d6bb7fdb0053c2218435e7e9e5/apps/test-suite/tests/Fetch.ts#L150-L172

**Reply by sudo2801** [reply]
@Kudo What is the expected release date? We're eagerly awaiting it!

**Reply by Kudo** [reply]
@sudo2801 it's available now in sdk 52 https://expo.dev/changelog/2024/11-12-sdk-52

**Reply by fukuli053** [reply]
@Kudo I tried the example provided in the documentation but wasn't successful. I'm sharing the code as an example. To understand where I got stuck, I added logging and didn't receive any logs beyond 'Fetching stream started'

```
const fetchStream = async () => {
      console.log("Fetching stream started");
      const resp = await fetch(
        "https://httpbin.org/drip?numbytes=512&duration=2",
        {
          headers: { Accept: "text/event-stream" },
        }
      );
      console.log("LOG: Response:", resp);
      const reader = resp.body?.getReader();
      console.log("LOG: Reader:", reader);
      const chunks = [];
      console.log("LOG: Chunks:", chunks);

      while (true) {
        const { done, value } = await reader.read();
        if (done) {
          console.log("Stream complete");
          break;
        }
        console.log("Received chunk:", value);
        chunks.push(value);
      }

      const buffer = new Uint8Array(chunks.reduce((acc, chunk) => acc + chunk.length, 0));
      console.log(buffer.length); // 512
    };

    fetchStream().catch((error) => {
      console.error("Error in fetchStream:", error);
    });
```

**Reply by Kudo** [reply]
@fukuli053 did you use the fetch imported from expo/fetch? `import { fetch } from 'expo/fetch';` we didn't overwrite the react-native's default fetch, so you need to import explicitly.

**Reply by fukuli053** [reply]
@Kudo yes, imported.

**Reply by Kudo** [reply]
i can see some `Received chunk` log on ios expo-go. did i miss something from the repro?

**Reply by fukuli053** [reply]
@Kudo I tried this repo (https://github.com/fukuli053/expo52-fetch-issue) on iOS 18 simulator with prebuild and expo go

**Reply by Kudo** [reply]
@fukuli053 it was my bad to have a regression since last week. https://github.com/expo/expo/pull/32894 should resolve the problem. will go for the pr review soon. thanks for catching the problem!

**Reply by Kudo** [reply]
@fukuli053 should work from `expo@52.0.7`. thanks again for catching the issue

**Reply by sudo2801** [reply]
@Kudo 

sometime my app crashes when I calling the fetch.

```
java.io.EOFException: source exhausted prematurely
  okio.InflaterSource.read(InflaterSource.kt:52)
  okio.GzipSource.read(GzipSource.kt:69)
  okio.RealBufferedSource.exhausted(RealBufferedSource.kt:197)
  expo.modules.fetch.NativeResponse.pumpResponseBodyStream(NativeResponse.kt:161)
  expo.modules.fetch.NativeResponse.access$pumpResponseBodyStream(NativeResponse.kt:17)
  expo.modules.fetch.NativeResponse$onResponse$1.invokeSuspend(NativeResponse.kt:119)
  kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
  kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:108)
  kotlinx.coroutines.internal.LimitedDispatcher$Worker.run(LimitedDispatcher.kt:115)
  kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:103)
  kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:584)
  kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:793)
  kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:697)
  kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:684)
```
